libcontainer/container_linux.go

```diff
@@ -1255,6 +1255,13 @@ func (c *linuxContainer) Restore(process *Process, criuOpts *CriuOpts) error {
 	for _, m := range c.config.Mounts {
 		switch m.Device {
 		case "bind":
+			// When runc starts a container it creates all mountpoints
+			// for bind mounts. Let's also create those mountpoints
+			// on restore to be able to restore a container created
+			// from an unmodified container rootfs.
+			if err := prepareBindMounts(m, c.config.Rootfs); err != nil {
```

Comment:
The same problem exists for tmpfs mounts which are specified in a container config. And there can be dependencies between mounts. We can mount tmpfs into /mnt and than bind-mount a host directory into /mnt/bind. This code will try to create /mnt/bind on a container rootfs and it will fail of rootfs is read-only. 

The only right way to handle this is to snapshot a container rootfs.

Reply:
But runc will also fail creating a container if the rootfs is read-only.

Reply:
So at least for a read-only rootfs container we should to the same steps during restore as during creation. It does not really make sense to take a snapshot of a read-only rootfs container as it is expected to be the same, but runc will still try to modify it.

Reply:
@adrianreber I don't understand me. Here are steps for creating a container:
1. createIfNotExists(/mnt). A container rootfs has /mnt, so this function does nothing and exits without errors
2. mount tmpfs into /mnt
3. createIfNotExists(/mnt/A) - create a directory A on a tmpfs mount
4. bind-mount a host path to /mnt/A

In case of restore, /mnt/A will be created on a container rootfs, what looks wrong, doesn't it?

Reply:
> In case of restore, /mnt/A will be created on a container rootfs, what looks wrong, doesn't it?

Not sure I understand you. Is this during restore or initial container creation?

If it is a tmpfs CRIU will anyway untar the content from the checkpoint into the tmpfs, so all tmpfs content will be handled correctly by CRIU internally.

If running the busybox container using podman, the downloaded container image (docker://busybox:latest) will not contain the directory /proc. runc will create /proc during container startup and mount proc at /proc.

If I checkpoint this container CRIU will remember that it mounted proc at /proc and try to restore the mount and fail.

I think there is a difference between changes done to the roof filesystem from the programs running in the container and changes done to the container by runc during startup.

Changes done to the container during startup, like creating /proc are necessary to restore the container.

I think restoring a container which does not modify the filesystem (or only using tmpfs) is a valid use case and because runc modifies the rootfs during startup we need to do the same modifications during restore. It seems unnecessary to have to transfer the changes done to the rootfs by runc during startup.

If the container (from within) modifies the root filesystem I agree we need to do a commit and export these changes to the new container. As long as the changes are done by runc, I think runc needs to handle the restore.

Reply:
> Not sure I understand you. Is this during restore or initial container creation?

This is an initial container creation.  I just show you that when runsc start a continaer, /mnt/A is created on a tmpfs mount, but on restore (with these changes) /mnt/A will be created on a container rootfs. You are right, criu will restore the tmpfs mount, but it will be later.

Reply:
Ah, okay, I get it. If a bind mount mountpoint is on a tmpfs runc will now modify the underlying rootfs, before CRIU mounts the tmpfs, understood. So skipping bind mount mountpoints would be useful, correct. Need to check what I need to do, to do that correctly.

Do you see any other problems if I can actually do the bind mount mountpoint on tmpfs correctly?

Reply:
I don't understand why you don't want to save a commit of  the highest layer of rootfs together with criu dump files. If we merge these changes, it will fix your test, but the problem will not disappear for cases when a container is started with a read-write rootfs and processes create files on it.

Reply:
Hmm... You are right, if the container modifies the rootfs the highest layer of the rootfs needs to be part of the checkpoint. What I am unsure about is the changes done to the rootfs by runc.

The correct thing to do would be that runc does not modify the rootfs at all. This, however, cannot be changed as it would break a lot of containers. This seems to be the root of this discussion.

My point is, that if runc changes the rootfs during creation (especially if it is a read-only container) then the same modification should happen during restore. Without this it will not be possible to migrate a container with a read-only rootfs, because runc will change read-only rootfs containers. Having to include the highest layer of a read-only rootfs container also seems wrong.

I understand your point of view and I would agree if runc would not modify the rootfs. Because runc modifies the rootfs during creation I think it can also be expected during restore.

Not sure how to continue this discussion. Maybe some comments from the runc maintainers would be helpful. How they see the changes to the rootfs done by runc. If these rootfs changes should also happen during restore or not. I guess my biggest problem with this is, is that runc does changes to the container, if this would not be case we would not need this whole discussion.

Reply:
> Do you see any other problems

Now you enumerate only bind-mounts, /proc and /sys, but I think you need to create destination directories for tmpfs mounts too.

Reply:
> > Do you see any other problems
> 
> Now you enumerate only bind-mounts, /proc and /sys, but I think you need to create destination directories for tmpfs mounts too.

Yes, just tested it and it failed. Will update this PR.